README.md
# pathlib_extensions
Backported and experimental filesystem path features for Python


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.2.1] - 2022-03-16

### Fixed

- Fixed bug where image, audio, or video files with upper case file extension are not recognized as such

## [0.2.0] - 2022-01-17

### Added

- `Path.is_image_file()`
- `Path.is_audio_file()`
- `Path.is_video_file()`

## [0.1.0] - 2022-01-11

### Added

- `PurePath.with_stem()`
- `PurePath.is_relative_to()`
- `Path.readlink()`
- `Path.link_to()`
- `Path.is_mount()`

test_pathlib_extensions.py
import os
import unittest
from test import test_pathlib, support

import pathlib_extensions


#
# Tests for the pure classes.
#

class _BasePurePathTest(test_pathlib._BasePurePathTest):
    def test_repr_common(self):
        for pathstr in ('a', 'a/b', 'a/b/c', '/', '/a/b', '/a/b/c'):
            p = self.cls(pathstr)
            clsname = p.__class__.__name__
            r = repr(p)
            # The repr() is in the form ClassName("forward-slashes path")
            self.assertTrue(r.startswith(clsname + '('), r)
            self.assertTrue(r.endswith(')'), r)
            inner = r[len(clsname) + 1: -1]
            self.assertEqual(eval(inner), p.as_posix())
            # The repr() roundtrips
            q = eval(r, pathlib_extensions.__dict__)
            self.assertIs(q.__class__, p.__class__)
            self.assertEqual(q, p)
            self.assertEqual(repr(q), r)

    # 3.9+
    if hasattr(test_pathlib._BasePurePathTest, "test_with_stem_common"):
        test_with_stem_common = test_pathlib._BasePurePathTest.test_with_stem_common
    # 3.6-3.8
    else:
        def test_with_stem_common(self):
            P = self.cls
            self.assertEqual(P('a/b').with_stem('d'), P('a/d'))
            self.assertEqual(P('/a/b').with_stem('d'), P('/a/d'))
            self.assertEqual(P('a/b.py').with_stem('d'), P('a/d.py'))
            self.assertEqual(P('/a/b.py').with_stem('d'), P('/a/d.py'))
            self.assertEqual(P('/a/b.tar.gz').with_stem('d'), P('/a/d.gz'))
            self.assertEqual(P('a/Dot ending.').with_stem('d'), P('a/d'))
            self.assertEqual(P('/a/Dot ending.').with_stem('d'), P('/a/d'))
            self.assertRaises(ValueError, P('').with_stem, 'd')
            self.assertRaises(ValueError, P('.').with_stem, 'd')
            self.assertRaises(ValueError, P('/').with_stem, 'd')
            self.assertRaises(ValueError, P('a/b').with_stem, '')
            self.assertRaises(ValueError, P('a/b').with_stem, '/c')
            self.assertRaises(ValueError, P('a/b').with_stem, 'c/')
            self.assertRaises(ValueError, P('a/b').with_stem, 'c/d')

    # 3.9+
    if hasattr(test_pathlib._BasePurePathTest, "test_is_relative_to_common"):
        test_is_relative_to_common = test_pathlib._BasePurePathTest.test_is_relative_to_common
    # 3.6-3.8
    else:
        def test_is_relative_to_common(self):
            P = self.cls
            p = P('a/b')
            self.assertRaises(TypeError, p.is_relative_to)
            self.assertRaises(TypeError, p.is_relative_to, b'a')
            self.assertTrue(p.is_relative_to(P()))
            self.assertTrue(p.is_relative_to(''))
            self.assertTrue(p.is_relative_to(P('a')))
            self.assertTrue(p.is_relative_to('a/'))
            self.assertTrue(p.is_relative_to(P('a/b')))
            self.assertTrue(p.is_relative_to('a/b'))
            # With several args.
            self.assertTrue(p.is_relative_to('a', 'b'))
            # Unrelated paths.
            self.assertFalse(p.is_relative_to(P('c')))
            self.assertFalse(p.is_relative_to(P('a/b/c')))
            self.assertFalse(p.is_relative_to(P('a/c')))
            self.assertFalse(p.is_relative_to(P('/a')))
            p = P('/a/b')
            self.assertTrue(p.is_relative_to(P('/')))
            self.assertTrue(p.is_relative_to('/'))
            self.assertTrue(p.is_relative_to(P('/a')))
            self.assertTrue(p.is_relative_to('/a'))
            self.assertTrue(p.is_relative_to('/a/'))
            self.assertTrue(p.is_relative_to(P('/a/b')))
            self.assertTrue(p.is_relative_to('/a/b'))
            # Unrelated paths.
            self.assertFalse(p.is_relative_to(P('/c')))
            self.assertFalse(p.is_relative_to(P('/a/b/c')))
            self.assertFalse(p.is_relative_to(P('/a/c')))
            self.assertFalse(p.is_relative_to(P()))
            self.assertFalse(p.is_relative_to(''))
            self.assertFalse(p.is_relative_to(P('a')))


class PurePosixPathTest(_BasePurePathTest, unittest.TestCase):
    cls = pathlib_extensions.PurePosixPath


class PureWindowsPathTest(_BasePurePathTest, unittest.TestCase):
    cls = pathlib_extensions.PureWindowsPath

    # 3.9+
    if hasattr(test_pathlib.PureWindowsPathTest, "test_with_stem"):
        test_with_stem = test_pathlib.PureWindowsPathTest.test_with_stem
    # 3.6-3.8
    else:
        def test_with_stem(self):
            P = self.cls
            self.assertEqual(P('c:a/b').with_stem('d'), P('c:a/d'))
            self.assertEqual(P('c:/a/b').with_stem('d'), P('c:/a/d'))
            self.assertEqual(P('c:a/Dot ending.').with_stem('d'), P('c:a/d'))
            self.assertEqual(P('c:/a/Dot ending.').with_stem('d'), P('c:/a/d'))
            self.assertRaises(ValueError, P('c:').with_stem, 'd')
            self.assertRaises(ValueError, P('c:/').with_stem, 'd')
            self.assertRaises(ValueError, P('//My/Share').with_stem, 'd')
            self.assertRaises(ValueError, P('c:a/b').with_stem, 'd:')
            self.assertRaises(ValueError, P('c:a/b').with_stem, 'd:e')
            self.assertRaises(ValueError, P('c:a/b').with_stem, 'd:/e')
            self.assertRaises(ValueError, P('c:a/b').with_stem, '//My/Share')

    # 3.9+
    if hasattr(test_pathlib.PureWindowsPathTest, "test_is_relative_to"):
        test_is_relative_to = test_pathlib.PureWindowsPathTest.test_is_relative_to
    # 3.6-3.8
    else:
        def test_is_relative_to(self):
            P = self.cls
            p = P('C:Foo/Bar')
            self.assertTrue(p.is_relative_to(P('c:')))
            self.assertTrue(p.is_relative_to('c:'))
            self.assertTrue(p.is_relative_to(P('c:foO')))
            self.assertTrue(p.is_relative_to('c:foO'))
            self.assertTrue(p.is_relative_to('c:foO/'))
            self.assertTrue(p.is_relative_to(P('c:foO/baR')))
            self.assertTrue(p.is_relative_to('c:foO/baR'))
            # Unrelated paths.
            self.assertFalse(p.is_relative_to(P()))
            self.assertFalse(p.is_relative_to(''))
            self.assertFalse(p.is_relative_to(P('d:')))
            self.assertFalse(p.is_relative_to(P('/')))
            self.assertFalse(p.is_relative_to(P('Foo')))
            self.assertFalse(p.is_relative_to(P('/Foo')))
            self.assertFalse(p.is_relative_to(P('C:/Foo')))
            self.assertFalse(p.is_relative_to(P('C:Foo/Bar/Baz')))
            self.assertFalse(p.is_relative_to(P('C:Foo/Baz')))
            p = P('C:/Foo/Bar')
            self.assertTrue(p.is_relative_to('c:'))
            self.assertTrue(p.is_relative_to(P('c:/')))
            self.assertTrue(p.is_relative_to(P('c:/foO')))
            self.assertTrue(p.is_relative_to('c:/foO/'))
            self.assertTrue(p.is_relative_to(P('c:/foO/baR')))
            self.assertTrue(p.is_relative_to('c:/foO/baR'))
            # Unrelated paths.
            self.assertFalse(p.is_relative_to(P('C:/Baz')))
            self.assertFalse(p.is_relative_to(P('C:/Foo/Bar/Baz')))
            self.assertFalse(p.is_relative_to(P('C:/Foo/Baz')))
            self.assertFalse(p.is_relative_to(P('C:Foo')))
            self.assertFalse(p.is_relative_to(P('d:')))
            self.assertFalse(p.is_relative_to(P('d:/')))
            self.assertFalse(p.is_relative_to(P('/')))
            self.assertFalse(p.is_relative_to(P('/Foo')))
            self.assertFalse(p.is_relative_to(P('//C/Foo')))
            # UNC paths.
            p = P('//Server/Share/Foo/Bar')
            self.assertTrue(p.is_relative_to(P('//sErver/sHare')))
            self.assertTrue(p.is_relative_to('//sErver/sHare'))
            self.assertTrue(p.is_relative_to('//sErver/sHare/'))
            self.assertTrue(p.is_relative_to(P('//sErver/sHare/Foo')))
            self.assertTrue(p.is_relative_to('//sErver/sHare/Foo'))
            self.assertTrue(p.is_relative_to('//sErver/sHare/Foo/'))
            self.assertTrue(p.is_relative_to(P('//sErver/sHare/Foo/Bar')))
            self.assertTrue(p.is_relative_to('//sErver/sHare/Foo/Bar'))
            # Unrelated paths.
            self.assertFalse(p.is_relative_to(P('/Server/Share/Foo')))
            self.assertFalse(p.is_relative_to(P('c:/Server/Share/Foo')))
            self.assertFalse(p.is_relative_to(P('//z/Share/Foo')))
            self.assertFalse(p.is_relative_to(P('//Server/z/Foo')))


class PurePathTest(_BasePurePathTest, unittest.TestCase):
    cls = pathlib_extensions.PurePath


#
# Tests for the concrete classes.
#

# Make sure any symbolic links in the base test path are resolved.
BASE = test_pathlib.BASE
join = test_pathlib.join
rel_join = test_pathlib.rel_join

only_nt = test_pathlib.only_nt
only_posix = test_pathlib.only_posix


@only_posix
class PosixPathAsPureTest(PurePosixPathTest):
    cls = pathlib_extensions.PosixPath


@only_nt
class WindowsPathAsPureTest(PureWindowsPathTest):
    cls = pathlib_extensions.WindowsPath


class _BasePathTest(test_pathlib._BasePathTest):
    """Tests for the FS-accessing functionalities of the Path classes."""

    # 3.9+
    if hasattr(test_pathlib._BasePathTest, "test_readlink"):
        test_readlink = test_pathlib._BasePathTest.test_readlink
    # 3.6-3.8
    else:
        @support.skip_unless_symlink
        def test_readlink(self):
            P = self.cls(BASE)
            self.assertEqual((P / 'linkA').readlink(), self.cls('fileA'))
            self.assertEqual((P / 'brokenLink').readlink(),
                             self.cls('non-existing'))
            self.assertEqual((P / 'linkB').readlink(), self.cls('dirB'))
            with self.assertRaises(OSError):
                (P / 'fileA').readlink()

    # 3.8+
    if hasattr(test_pathlib._BasePathTest, "test_link_to"):
        test_link_to = test_pathlib._BasePathTest.test_link_to
    # 3.6-3.7
    else:
        @unittest.skipUnless(hasattr(os, "link"), "os.link() is not present")
        def test_link_to(self):
            P = self.cls(BASE)
            p = P / 'fileA'
            size = p.stat().st_size
            # linking to another path.
            q = P / 'dirA' / 'fileAA'
            try:
                p.link_to(q)
            except PermissionError as e:
                self.skipTest('os.link(): %s' % e)
            self.assertEqual(q.stat().st_size, size)
            self.assertEqual(os.path.samefile(p, q), True)
            self.assertTrue(p.stat)
            # Linking to a str of a relative path.
            r = rel_join('fileAAA')
            q.link_to(r)
            self.assertEqual(os.stat(r).st_size, size)
            self.assertTrue(q.stat)

    if hasattr(test_pathlib._BasePathTest, "test_link_to_not_implemented"):
        test_link_to_not_implemented = test_pathlib._BasePathTest.test_link_to_not_implemented
    # 3.6-3.7
    else:
        @unittest.skipIf(hasattr(os, "link"), "os.link() is present")
        def test_link_to_not_implemented(self):
            P = self.cls(BASE)
            p = P / 'fileA'
            # linking to another path.
            q = P / 'dirA' / 'fileAA'
            with self.assertRaises(NotImplementedError):
                p.link_to(q)

    # 3.7+
    if hasattr(test_pathlib._BasePathTest, "test_is_mount"):
        test_is_mount = test_pathlib._BasePathTest.test_is_mount
    # 3.6
    else:
        @only_posix
        def test_is_mount(self):
            P = self.cls(BASE)
            R = self.cls('/')  # TODO: Work out Windows.
            self.assertFalse((P / 'fileA').is_mount())
            self.assertFalse((P / 'dirA').is_mount())
            self.assertFalse((P / 'non-existing').is_mount())
            self.assertFalse((P / 'fileA' / 'bah').is_mount())
            self.assertTrue(R.is_mount())
            if support.can_symlink():
                self.assertFalse((P / 'linkA').is_mount())
            self.assertIs(self.cls('/\udfff').is_mount(), False)
            self.assertIs(self.cls('/\x00').is_mount(), False)


class PathTest(_BasePathTest, unittest.TestCase):
    cls = pathlib_extensions.Path


@only_posix
class PosixPathTest(_BasePathTest, unittest.TestCase):
    cls = pathlib_extensions.PosixPath


@only_nt
class WindowsPathTest(_BasePathTest, unittest.TestCase):
    cls = pathlib_extensions.WindowsPath


if __name__ == '__main__':
    unittest.main()


src/pathlib_extensions/pathlib_extensions.py
import os
import pathlib

from mimetypes_extensions import image_file_extensions, audio_file_extensions, video_file_extensions

__all__ = [
    "PurePath", "PurePosixPath", "PureWindowsPath",
    "Path", "PosixPath", "WindowsPath",
]

#
# Internals
#

if hasattr(pathlib, "_IGNORED_ERROS"):
    _IGNORED_ERROS = pathlib._IGNORED_ERROS
else:
    from errno import ENOENT, ENOTDIR, EBADF, ELOOP

    # EBADF - guard against macOS `stat` throwing EBADF
    _IGNORED_ERROS = (ENOENT, ENOTDIR, EBADF, ELOOP)

if hasattr(pathlib, "_IGNORED_WINERRORS"):
    _IGNORED_WINERRORS = pathlib._IGNORED_WINERRORS
else:
    _IGNORED_WINERRORS = (
        21,  # ERROR_NOT_READY - drive exists but is not accessible
        123,  # ERROR_INVALID_NAME - fix for bpo-35306
        1921,  # ERROR_CANT_RESOLVE_FILENAME - fix for broken symlink pointing to itself
    )

if hasattr(pathlib, "_ignore_error"):
    _ignore_error = pathlib._ignore_error
else:
    def _ignore_error(exception):
        return (getattr(exception, 'errno', None) in _IGNORED_ERROS or
                getattr(exception, 'winerror', None) in _IGNORED_WINERRORS)


class _NormalAccessor(pathlib._NormalAccessor):
    if hasattr(os, "link"):
        link_to = os.link
    else:
        @staticmethod
        def link_to(self, target):
            raise NotImplementedError("os.link() not available on this system")


_normal_accessor = _NormalAccessor()


#
# Public API
#

class PurePath(pathlib.PurePath):
    """Base class for manipulating paths without I/O.

    PurePath represents a filesystem path and offers operations which
    don't imply any actual filesystem I/O.  Depending on your system,
    instantiating a PurePath will return either a PurePosixPath or a
    PureWindowsPath object.  You can also instantiate either of these classes
    directly, regardless of your system.
    """

    def __new__(cls, *args):
        """Construct a PurePath from one or several strings and or existing
        PurePath objects.  The strings and path objects are combined so as
        to yield a canonicalized path, which is incorporated into the
        new PurePath object.
        """
        if cls is PurePath:
            cls = PureWindowsPath if os.name == 'nt' else PurePosixPath
        return cls._from_parts(args)

    # 3.9+
    if hasattr(pathlib.PurePath, "with_stem"):
        with_stem = pathlib.PurePath.with_stem
    # 3.6-3.8
    else:
        def with_stem(self, stem):
            """Return a new path with the stem changed."""
            return self.with_name(stem + self.suffix)

    # 3.9+
    if hasattr(pathlib.PurePath, "is_relative_to"):
        is_relative_to = pathlib.PurePath.is_relative_to
    # 3.6-3.8
    else:
        def is_relative_to(self, *other):
            """Return True if the path is relative to another path or False.
            """
            try:
                self.relative_to(*other)
                return True
            except ValueError:
                return False


# Can't subclass os.PathLike from PurePath and keep the constructor
# optimizations in PurePath._parse_args().
os.PathLike.register(PurePath)


class PurePosixPath(PurePath):
    """PurePath subclass for non-Windows systems.

    On a POSIX system, instantiating a PurePath should return this object.
    However, you can also instantiate it directly on any system.
    """
    _flavour = pathlib._posix_flavour
    __slots__ = ()


class PureWindowsPath(PurePath):
    """PurePath subclass for Windows systems.

    On a Windows system, instantiating a PurePath should return this object.
    However, you can also instantiate it directly on any system.
    """
    _flavour = pathlib._windows_flavour
    __slots__ = ()


# Filesystem-accessing classes


class Path(pathlib.Path):
    """PurePath subclass that can make system calls.

    Path represents a filesystem path but unlike PurePath, also offers
    methods to do system calls on path objects. Depending on your system,
    instantiating a Path will return either a PosixPath or a WindowsPath
    object. You can also instantiate a PosixPath or WindowsPath directly,
    but cannot instantiate a WindowsPath on a POSIX system or vice versa.
    """

    def __new__(cls, *args, **kwargs):
        if cls is Path:
            cls = WindowsPath if os.name == 'nt' else PosixPath
        self = cls._from_parts(args, init=False)
        if not self._flavour.is_supported:
            raise NotImplementedError("cannot instantiate %r on your system"
                                      % (cls.__name__,))
        self._init()
        return self

    def _init(self,
              # Private non-constructor arguments
              template=None,
              ):
        self._closed = False
        if template is not None:
            self._accessor = template._accessor
        else:
            self._accessor = _normal_accessor

    # Public API

    # 3.9+
    if hasattr(pathlib.Path, "readlink"):
        readlink = pathlib.Path.readlink
    # 3.6-3.8
    else:
        def readlink(self):
            """
            Return the path to which the symbolic link points.
            """
            path = self._accessor.readlink(self)
            obj = self._from_parts((path,), init=False)
            obj._init(template=self)
            return obj

    # 3.8+
    if hasattr(pathlib.Path, "link_to"):
        link_to = pathlib.Path.link_to
    # 3.6-3.7
    else:
        def link_to(self, target):
            """
            Make the target path a hard link pointing to this path.

            Note this function does not make this path a hard link to *target*,
            despite the implication of the function and argument names. The order
            of arguments (target, link) is the reverse of Path.symlink_to, but
            matches that of os.link.
            """
            if self._closed:
                self._raise_closed()
            self._accessor.link_to(self, target)

    # Convenience functions for querying the stat results

    # 3.7+
    if hasattr(pathlib.Path, "is_mount"):
        is_mount = pathlib.Path.is_mount
    # 3.6
    else:
        def is_mount(self):
            """
            Check if this path is a POSIX mount point
            """
            # Need to exist and be a dir
            if not self.exists() or not self.is_dir():
                return False

            parent = Path(self.parent)
            try:
                parent_dev = parent.stat().st_dev
            except OSError:
                return False

            dev = self.stat().st_dev
            if dev != parent_dev:
                return True
            ino = self.stat().st_ino
            parent_ino = parent.stat().st_ino
            return ino == parent_ino

    # Experimental functions

    def is_image_file(self):
        """
        Whether this path is an image file.
        """
        try:
            return self.is_file() and (self.suffix.lower() in image_file_extensions)
        except OSError as e:
            if not _ignore_error(e):
                raise
            # Path doesn't exist or is a broken symlink
            return False
        except ValueError:
            # Non-encodable path
            return False

    def is_audio_file(self):
        """
        Whether this path is an audio file.
        """
        try:
            return self.is_file() and (self.suffix.lower() in audio_file_extensions)
        except OSError as e:
            if not _ignore_error(e):
                raise
            # Path doesn't exist or is a broken symlink
            return False
        except ValueError:
            # Non-encodable path
            return False

    def is_video_file(self):
        """
        Whether this path is a video file.
        """
        try:
            return self.is_file() and (self.suffix.lower() in video_file_extensions)
        except OSError as e:
            if not _ignore_error(e):
                raise
            # Path doesn't exist or is a broken symlink
            return False
        except ValueError:
            # Non-encodable path
            return False


class PosixPath(Path, PurePosixPath):
    """Path subclass for non-Windows systems.

    On a POSIX system, instantiating a Path should return this object.
    """
    __slots__ = ()


class WindowsPath(Path, PureWindowsPath):
    """Path subclass for Windows systems.

    On a Windows system, instantiating a Path should return this object.
    """
    __slots__ = ()

    def owner(self):
        raise NotImplementedError("Path.owner() is unsupported on this system")

    def group(self):
        raise NotImplementedError("Path.group() is unsupported on this system")

    # 3.7+
    if hasattr(pathlib.WindowsPath, "is_mount"):
        is_mount = pathlib.WindowsPath.is_mount
    # 3.6
    else:
        def is_mount(self):
            raise NotImplementedError("Path.is_mount() is unsupported on this system")
